Update version for smart-meter-texas to 0.4.7

meyerrj opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Update library version to fix a broken API call.
Changelog: https://github.com/grahamwetzler/smart-meter-texas/compare/v0.4.0...v0.4.7

Potential fix for #40348


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ x] Dependency upgrade
- [ x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #40348
- This PR is related to issue: #40348

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

meyerrj committed:
Update version for smart-meter-texas to 0.4.4

Potential fix for https://github.com/home-assistant/core/issues/40348

homeassistant commented:
Hi @meyerrj, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#54493).

Once you do that we will be able to review and accept this pull request.

Thanks!

grahamwetzler commented:
Thanks again @meyerrj, I just deployed v0.4.4 https://pypi.org/manage/project/smart-meter-texas/release/0.4.4/

meyerrj commented:
No problem @grahamwetzler! Thanks for all your work on this project!

bdraco commented:
You'll need to regenerate requirements

`python3 -m script.gen_requirements_all`

bdraco reviewed (changes requested):
Please see above

meyerrj committed:
Fix requirements for smart_meter_texas

bdraco commented:
Seeing this when testing


```
2021-08-12 03:15:59 ERROR (MainThread) [homeassistant.components.smart_meter_texas.config_flow] Cannot connect
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 969, in _wrap_create_connection
    return await self._loop.create_connection(*args, **kwargs)  # type: ignore  # noqa
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1050, in create_connection
    transport, protocol = await self._create_connection_transport(
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1080, in _create_connection_transport
    await waiter
  File "/usr/local/lib/python3.8/asyncio/sslproto.py", line 529, in data_received
    ssldata, appdata = self._sslpipe.feed_ssldata(data)
  File "/usr/local/lib/python3.8/asyncio/sslproto.py", line 189, in feed_ssldata
    self._sslobj.do_handshake()
  File "/usr/local/lib/python3.8/ssl.py", line 944, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1125)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/smart_meter_texas/config_flow.py", line 37, in validate_input
    await client.authenticate()
  File "/usr/local/lib/python3.8/site-packages/smart_meter_texas/__init__.py", line 166, in authenticate
    await self._init_websession()
  File "/usr/local/lib/python3.8/site-packages/smart_meter_texas/__init__.py", line 126, in _init_websession
    await self.websession.get(BASE_URL, headers=self._agent_headers())
  File "/usr/local/lib/python3.8/site-packages/aiohttp/client.py", line 520, in _request
    conn = await self._connector.connect(
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 535, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 892, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 1051, in _create_direct_connection
    raise last_exc
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 1020, in _create_direct_connection
    transp, proto = await self._wrap_create_connection(
  File "/usr/local/lib/python3.8/site-packages/aiohttp/connector.py", line 971, in _wrap_create_connection
    raise ClientConnectorCertificateError(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorCertificateError: Cannot connect to host smartmetertexas.com:443 ssl:True [SSLCertVerificationError: (1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1125)')]

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/smart_meter_texas/config_flow.py", line 58, in async_step_user
    info = await validate_input(self.hass, user_input)
  File "/usr/src/homeassistant/homeassistant/components/smart_meter_texas/config_flow.py", line 39, in validate_input
    raise CannotConnect from error
homeassistant.components.smart_meter_texas.config_flow.CannotConnect

```

meyerrj commented:
Could the exception possibly be an issue with the SSL library on the client?

meyerrj commented:
I found these issues:  
https://bugs.python.org/issue36137  
https://bugs.python.org/issue36011 

It appears that Python's default SSL library on various platforms may be missing some CAs needed to verify some websites, It has some interesting behavior when working through Docker on Windows too.

meyerrj committed:
Merge branch 'home-assistant:dev' into dev

meyerrj commented:
@bdraco  Just wanted to follow up. Can you confirm that the SSL error you received wasn't an issue with your Python/SSL environment?

bdraco commented:
I tried production and dev. Error both places

meyerrj commented:
@bdraco  Everything I can find on the error points to an issue with the local certificates being out of date or incomplete...

Short of making the library ignore SSL validation (which isn't recommended), I think this issue needs to be resolved in the environment somehow. 

@grahamwetzler  Another thought may be to import the certifi module and use that ssl context... https://github.com/aio-libs/aiohttp/issues/955#issuecomment-568425975   Thoughts?

meyerrj commented:
I found an issue in the site's certificate chain using ssllabs.com to test the site... 

I'm now able to reproduce the error with the library itself, so back to testing for a solution. 

![image](https://user-images.githubusercontent.com/5955723/129427037-e3b555bb-7115-4b05-a2d8-74e45db8db82.png)

![image](https://user-images.githubusercontent.com/5955723/129426983-06eb94f6-acdc-4ed2-bb5f-a785addbabe6.png)

bdraco commented:
Is it a new intermediate certificate that isn't in the version of certifi that home assistant uses?

Does upgrading to certifi 2021.5.30 include it?

meyerrj commented:
So, this definitely appears to get even more troublesome than I had anticipated...  

The library works with Python on Windows for me just fine, but not Linux.  Updating certifi did not help, either. 

The only way I can currently get it to work on Linux is by setting the sslcontext.verify_mode to none, or by passing in the certificate chain file when working against smartmetertexas.com. 

Also, the SSL certificate chain is correct for the www.smartmetertexas.com domain, but entirely different for smartmetertexas.com.  The ssllabs test returns different results too. 🤦‍♂️

![image](https://user-images.githubusercontent.com/5955723/129430566-f225bf26-1daf-4f31-a629-107efa9030ec.png)

meyerrj committed:
Merge branch 'home-assistant:dev' into dev

bdraco commented:
@meyerrj 

```
> wget http://cacerts.digicert.com/DigiCertSHA2SecureServerCA.crt
> openssl x509 -inform der -in DigiCertSHA2SecureServerCA.crt  -out  DigiCertSHA2SecureServerCA.pem
> python3
import asyncio
import certifi
import aiohttp
import ssl

ssl_context = ssl.create_default_context(cafile=certifi.where())
ssl_context.load_verify_locations('DigiCertSHA2SecureServerCA.pem')

async def _test_get():
	async with aiohttp.ClientSession() as session:
		return await session.get('https://smartmetertexas.com', ssl=ssl_context)

asyncio.run(_test_get())
```

If you want to make it future proof, connect and get the cert, follow the ca-issuers url, download the intermediate cert, and add it to the trust chain.

I just extracted the ca-issuers and manually wgetted it above.

meyerrj commented:
@bdraco Thanks for the lead! I'm still rusty with my python, so this helps a lot!

I currently have some logic that can at least check/compare the known fingerprint until it expires or changes, but this sounds like a better approach. I'll see if I can get something working in the library code.

Are there any file storage considerations I'd need to keep in mind with HA if the library were to download CA files for local use?

bdraco commented:
homeassistant.helpers.storage.Storage is likely your best bet

meyerrj committed:
Merge branch 'home-assistant:dev' into dev

meyerrj committed:
Merge branch 'home-assistant:dev' into dev

meyerrj committed:
SSL context improvements for Smart Meter Texas.

meyerrj commented:
Update:  I will hopefully have a pending change incoming soon that incorporates the logic above to retrieve the SSL certificate and load it into the SSL context. It will then pass that context into the smart meter library code and use that. 

This change assumes another version bump on the smart meter library, which isn't available yet.  If my changes are accepted and the version is bumped I can push my changes here and if everything goes well, we should have a working implementation again.  🤞

meyerrj committed:
Version bump for smart_meter_texas

meyerrj committed:
Cleanups.
Moved the ssl context logic into its own file to be imported.
Updated the config flow to account for the ssl context.

meyerrj committed:
Merge branch 'home-assistant:dev' into dev

meyerrj committed:
Merge branch 'dev' of github.com:meyerrj/core into dev

bdraco commented on homeassistant/components/smart_meter_texas/get_ssl_context.py:
```diff
@@ -0,0 +1,57 @@
+    ssl_context = None
+    try:
+        caiKey = "CA Issuers - URI:"
+        re_issuers_uri = re.compile(r"(https?://+[\w\d:#@%/;$()~_?\+-=\\\.&]*)", re.UNICODE)
```
This seem brittle as it relies on extracting undecoded ASN.1 data.  We should decode the data and get the url instead

meyerrj replied:
I tried a few different approaches but wasn't able to figure out how to just get the CA Issuers URI from the previously decoded asn1.

bdraco replied:
Using the asn.1 example, something like this might work:

```python3
import ssl, socket
import OpenSSL.crypto as crypto
import asn1
import sys


tag_id_to_string_map = {
    asn1.Numbers.Boolean: "BOOLEAN",
    asn1.Numbers.Integer: "INTEGER",
    asn1.Numbers.BitString: "BIT STRING",
    asn1.Numbers.OctetString: "OCTET STRING",
    asn1.Numbers.Null: "NULL",
    asn1.Numbers.ObjectIdentifier: "OBJECT",
    asn1.Numbers.PrintableString: "PRINTABLESTRING",
    asn1.Numbers.IA5String: "IA5STRING",
    asn1.Numbers.UTCTime: "UTCTIME",
    asn1.Numbers.Enumerated: "ENUMERATED",
    asn1.Numbers.Sequence: "SEQUENCE",
    asn1.Numbers.Set: "SET",
}


class_id_to_string_map = {
    asn1.Classes.Universal: "U",
    asn1.Classes.Application: "A",
    asn1.Classes.Context: "C",
    asn1.Classes.Private: "P",
}


def tag_id_to_string(identifier):
    """Return a string representation of a ASN.1 id."""
    if identifier in tag_id_to_string_map:
        return tag_id_to_string_map[identifier]
    return "{:#02x}".format(identifier)


def class_id_to_string(identifier):
    """Return a string representation of an ASN.1 class."""
    if identifier in class_id_to_string_map:
        return class_id_to_string_map[identifier]
    raise ValueError("Illegal class: {:#02x}".format(identifier))


def value_to_string(tag_number, value):
    if tag_number == asn1.Numbers.ObjectIdentifier:
        return value
    elif isinstance(value, bytes):
        return "0x" + str(binascii.hexlify(value).upper())
    elif isinstance(value, str):
        return value
    else:
        return repr(value)


def pretty_print(input_stream, output_stream, indent=0):
    """Pretty print ASN.1 data."""
    while not input_stream.eof():
        tag = input_stream.peek()
        if tag.typ == asn1.Types.Primitive:
            tag, value = input_stream.read()
            output_stream.write(" " * indent)
            output_stream.write(
                "[{}] {}: {}\n".format(
                    class_id_to_string(tag.cls),
                    tag_id_to_string(tag.nr),
                    value_to_string(tag.nr, value),
                )
            )
        elif tag.typ == asn1.Types.Constructed:
            output_stream.write(" " * indent)
            output_stream.write(
                "[{}] {}\n".format(
                    class_id_to_string(tag.cls), tag_id_to_string(tag.nr)
                )
            )
            input_stream.enter()
            pretty_print(input_stream, output_stream, indent + 2)
            input_stream.leave()


ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE
with ctx.wrap_socket(socket.socket(), server_hostname="smartmetertexas.com") as s:
    s.connect(("smartmetertexas.com", 443))
    cert_bin = s.getpeercert(True)
    x509 = crypto.load_certificate(crypto.FILETYPE_ASN1, cert_bin)
    for idx in range(x509.get_extension_count()):
        ext = x509.get_extension(idx)
        short_name = ext.get_short_name()
        if short_name == b"authorityInfoAccess":
            decoder = asn1.Decoder()
            print(str(ext))
            data = ext.get_data()
            decoder.start(data)
            pretty_print(decoder, sys.stdout)
```

meyerrj replied:
Thanks! I was able to get something working by narrowing it down to this:

```
def value_to_string(tag_number, value):
    if tag_number == asn1.Numbers.ObjectIdentifier:
        return value
    elif isinstance(value, bytes):
        return "0x" + str(binascii.hexlify(value).upper())
    elif isinstance(value, str):
        return value
    else:
        return repr(value)

def find_ca_issuers_uri(input_stream, tag_ca_issuers_uri_found = False):
    """Lookup the CA Issuers - URI Object and return the value."""
    ca_issuers_uri = None
    while not input_stream.eof() and not ca_issuers_uri:
        tag = input_stream.peek()
        if tag.typ == asn1.Types.Primitive:
            tag, value = input_stream.read()
            # If we've located the tag, extract the URI value
            if (tag_ca_issuers_uri_found):
                str_value = value_to_string(tag.nr, value)
                if (str_value):
                    ca_issuers_uri = str_value.decode('utf-8')
                    tag_ca_issuers_uri_found = False
                    break
                else:
                    tag_ca_issuers_uri_found = False
            # Locate the tag for CA Issuers - URI
            elif (value_to_string(tag.nr, value) == '1.3.6.1.5.5.7.48.2'):
                tag_ca_issuers_uri_found = True

        elif tag.typ == asn1.Types.Constructed:
            input_stream.enter()
            ca_issuers_uri = find_ca_issuers_uri(input_stream)
            input_stream.leave()

    return ca_issuers_uri

async def main():
    ca_issuers_uri = None
    context = ssl.create_default_context(capath=certifi.where())
    context.check_hostname = False
    context.verify_mode = ssl.CERT_NONE
    with context.wrap_socket(socket.socket(), server_hostname=BASE_HOSTNAME) as s:
        s.connect((BASE_HOSTNAME, 443))
        cert_bin = s.getpeercert(True)
        x509 = crypto.load_certificate(crypto.FILETYPE_ASN1, cert_bin)
        for idx in range(x509.get_extension_count()):
            ext = x509.get_extension(idx)
            short_name = ext.get_short_name()
            if short_name == b"authorityInfoAccess":
                decoder = asn1.Decoder()
                data = ext.get_data()
                decoder.start(data)
                ca_issuers_uri = find_ca_issuers_uri(decoder)

    print(ca_issuers_uri)

```

bdraco commented on homeassistant/components/smart_meter_texas/get_ssl_context.py:
```diff
@@ -0,0 +1,57 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+def get_ssl_context():
```
This needs to live in the pypi package per https://developers.home-assistant.io/docs/development_checklist/

meyerrj replied:
Would including the function within the smart_meter_texas library be acceptable?  
If so, @grahamwetzler  do you have an opinion on where this logic should live?

bdraco replied:
> Would including the function within the smart_meter_texas library be acceptable?

Yes 👍

grahamwetzler replied:
I think it’s fine to move into `smart_meter_texas/__init__.py` for now.

meyerrj replied:
Sounds good. 

I'll create a static class named ClientSSLContext and put the get_ssl_context method in there, along with any helper methods needed for getting the CA Issuer URI via the asn.1 data.  I'll work on getting the asyncio/executor logic working with this change as well.

I will keep it so the Client uses a standard SSL Context by default unless one is passed in explicitly during initialization. 

This sure is turning out to be a fun and educational exercise with python. 😅

I greatly appreciate all the help, feedback, and patience!

bdraco commented on homeassistant/components/smart_meter_texas/get_ssl_context.py:
```diff
@@ -0,0 +1,57 @@
+        ssl_context = ssl.create_default_context(capath=certifi.where())
+        ssl_context.check_hostname = False
+        ssl_context.verify_mode = ssl.CERT_NONE
+        with ssl_context.wrap_socket(socket.socket(), server_hostname=BASE_HOSTNAME) as s:
+            s.connect((BASE_HOSTNAME, 443))
+            cert_bin = s.getpeercert(True)
```
This is doing I/O in the event loop.  It needs to run in the executor or use asyncio.open_connection

meyerrj replied:
I'll see what I can do to make this work with asyncio.

bdraco replied:
https://docs.python.org/3/library/asyncio-protocol.html#asyncio.BaseTransport.get_extra_info

Looks like `get_extra_info` doesn't have an easy way to get the binary form so you might be stuck running it in the executor

meyerrj replied:
I think what I have in my current changes for the library should make the function call with the Socket connection run in the Executor... 

https://github.com/meyerrj/smart-meter-texas/blob/fc6d5fccafae59710b30db42d244778808c60796/smart_meter_texas/__init__.py#L361  

If you have a chance to review and let me know if you see anything that needs to be adjusted that would be helpful.

bdraco replied:
Looks like that should do it 👍

bdraco reviewed (changes requested):
Looks like this is getting closer 👍 

Please see comments above

meyerrj committed:
SSL updates

meyerrj committed:
Smart meter texas 0.4.7

meyerrj committed:
Merge branch 'home-assistant:dev' into dev

meyerrj commented:
Okay, fingers crossed, this latest push should do the trick... 😁

bdraco commented:
> Okay, fingers crossed, this latest push should do the trick... 😁

I started the CI.  Also for future PRs, please be sure to branch so the incoming branch isn't named `dev`.

bdraco commented:
Looks like one more tweak is needed

I tried creating a fresh config flow
```
2021-08-24 15:44:43 ERROR (MainThread) [homeassistant.components.smart_meter_texas.config_flow] Unexpected exception
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/smart_meter_texas/config_flow.py", line 58, in async_step_user
    info = await validate_input(self.hass, user_input)
  File "/usr/src/homeassistant/homeassistant/components/smart_meter_texas/config_flow.py", line 31, in validate_input
    ssl_context = await ClientSSLContext.get_ssl_context()
TypeError: get_ssl_context() missing 1 required positional argument: 'self'
```

meyerrj committed:
Update config_flow.py

Update call for ssl_context

meyerrj commented:
I'll be more mindful of the branch name next time. :)

Hopefully fixed the call on the config flow this time.

bdraco commented:
Config flow is working now. Same issue in `__init__.py`
```
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 304, in async_setup
    result = await component.async_setup_entry(hass, self)  # type: ignore
  File "/usr/src/homeassistant/homeassistant/components/smart_meter_texas/__init__.py", line 45, in async_setup_entry
    ssl_context = await ClientSSLContext.get_ssl_context()
TypeError: get_ssl_context() missing 1 required positional argument: 'self'
```

meyerrj committed:
Update __init__.py

Fix call for ClientSSLContext client_ssl_context

bdraco commented:
Its setting up now, but the meter is showing unavailable

meyerrj commented:
I noticed that the smart meter website can have some latency with getting your meter and meter stats on first setup, and sometimes it takes a while before it actually returns result for your "current" usage. Sometimes I'm only able to query the previous day's information until late in the day.

bdraco commented:
I started the CI.  I'll recheck it in 30m and see if it becomes available

bdraco commented:
It's working. Just too 6 minutes to do the first poll

meyerrj commented:
Hooray! Thanks for all the help in getting this resolved! I'm excited to get this working on my HA. 😁

bdraco reviewed (approved):
Should be good to go if the CI passes

bdraco commented:
If you want to add support for the energy integration in a future PR:

https://developers.home-assistant.io/docs/core/entity/sensor#long-term-statistics

bdraco commented:
Looks like there are some formatting and a pylint issue to address

meyerrj committed:
Updates using Black

meyerrj commented:
GitHub's web editor is not well suited for Python. 🤦‍♂️

bdraco commented on homeassistant/components/smart_meter_texas/__init__.py:
```diff
@@ -87,13 +92,18 @@ class SmartMeterTexasData:
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        account: Account,
+        ssl: ssl.SSLContext,
```
```suggestion
        ssl_context: ssl.SSLContext,
```

bdraco commented on homeassistant/components/smart_meter_texas/__init__.py:
```diff
@@ -87,13 +92,18 @@ class SmartMeterTexasData:
         """Initialize the data coordintator."""
         self._entry = entry
         self.account = account
+        self.ssl_context = ssl
```
```suggestion
```

bdraco commented on homeassistant/components/smart_meter_texas/__init__.py:
```diff
@@ -87,13 +92,18 @@ class SmartMeterTexasData:
+        self.ssl_context = ssl
         websession = aiohttp_client.async_get_clientsession(hass)
-        self.client = Client(websession, account)
+        self.client = Client(websession, account, ssl_context=self.ssl_context)
```
```suggestion
        self.client = Client(websession, account, ssl_context=ssl_context)
```

bdraco commented on homeassistant/components/smart_meter_texas/__init__.py:
```diff
@@ -1,8 +1,9 @@
 """The Smart Meter Texas integration."""
 import asyncio
 import logging
+import ssl
```
```suggestion
```

meyerrj committed:
Update homeassistant/components/smart_meter_texas/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

meyerrj committed:
Update homeassistant/components/smart_meter_texas/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

meyerrj committed:
Update homeassistant/components/smart_meter_texas/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

meyerrj committed:
Update homeassistant/components/smart_meter_texas/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented on homeassistant/components/smart_meter_texas/__init__.py:
```diff
@@ -2,7 +2,7 @@
 import asyncio
 import logging
```
```suggestion
import logging
import ssl
```

Looks like you still need it for the typing on `ssl.SSLContext`

meyerrj committed:
Update homeassistant/components/smart_meter_texas/__init__.py

Co-authored-by: J. Nick Koston <nick@koston.org>

bdraco commented:
Checked manually, should be good now.

Retesting...

bdraco commented:
testing looks good 👍

meyerrj commented:
Sorry for the trouble with this.  My python is admittedly a little rusty after roughly a decade. 
This has certainly been a fun deep dive to relearn some lessons.

bdraco commented:
CI failure is unrelated and being worked on in https://github.com/home-assistant/core/pull/55169
